design/open/kv-query-tool.md

```diff
@@ -14,32 +16,69 @@ We will elevate this logic and into a new command that will provide the ability
+
+### Full Knowledge
+
+With this approach, `lakefs kv` is implemented with all the knowledge available. Knowledge like: how to generate a repository partition from the repository ID and object, how to generate a branch key from the repository and branch IDs, the fact that staged objects' keys cannot be provided as strings etc.
+The commands in that case will be very specific, having the user specifying which entity he is working with. Essentially, this will be a CLI interface to various services, using KV (graveler, auth, etc.)
+* `lakefs kv get-repository <REPO_ID>`
+* `lakefs kv scan-staging <STAGING_TOKEN>` etc.
+With this approach, `lakefs kv` will not have functionalities such as `get-staged-object` by key, or any other operation we know is not possible or does not make any sense
 
-* `lakefs kv get <partition> <key>` </br>
-  Given a `partition key` and a `key` (both as strings), returns the corresponding object (json format)
+Pros:
+* Implementation is pretty straight forward - we actually already have everything implemented, and only need to make it accessible via `lakefs kv` tool
 
-* `lakefs kv scan <partition> [<key>]` </br>
-  Given a `partition key` and an optional `key`, returns all object in that partition that keys are equal or greater to the specified `key` (no `key` will return all objects in the partition). Results will be formatted as a json array
+Cons:
+* Lack of flexibility - adding a new entity, requires to implement new commands (bring in the knowledge) in the `lakefs kv` tool
+* Not really a "KV Querying" but more of a "services querying" tool. This might bring bugs from the existing logic to mask KV problems, or make them harder to identify
 
-* `lakefs kv set <partition> <key> <value>` </br>
-  Given a `partition key`, a `key` and a json formatted object, sets the value for the given `key` in the given `partition`, as an encoded protobuf for the given object
+### Minimal Knowledge
 
-### Delete (?)
-* `lakefs kv delete <partition> <key>` </br>
-  Given a `partition key` and a `key`, deletes the corresponding entry from the KV Store
-  **Question:** Do we really want this?
+With this approach, `lakefs kv` will have the minimal required knowledge of objects and their structures. It will present a generic KV-like interface that works with `partition`, `key` and `entries`.
+* `lakefs kv get <partition> <key>`
+* `lakefs kv scan <partition> [<key>]`
+This implementation will acquire the required data dynamically at runtime, by having entities register their a key parsing/matching function, an object structure and any other needed data, I cannot think of at the moment. This approach relies on the fact that the `key` structure "hides" information about the entity type, and so `lakefs kv` can leverage that in a sort of reflection-like way, to access the correct entity and construct the corresponding object, in order to decode the entry data correctly and  display it
```

Comment:
I don't think reflection is needed. Like you wrote, it's more of a regex matching keys problem, with each matcher linking to a protobuf schema.

Reply:
That was exactly my intention, hence the reflection-like